### DISCUSSION TITLE
Devbox vs. plain Nix

### DISCUSSION THREAD

**Discussion by sisp** [copier-org/copier#1468] [Ideas]
We're currently using Nix for two things:

1. Setting up reproducible dev environments
2. Providing deterministic runtime environments including non-Python dependencies like Git

Nix has a somewhat steep learning curve IMHO. I've stumbled over [Devbox](https://github.com/jetpack-io/devbox), which is an abstraction on top of Nix tailored to the dev environments use case, even with _granular_ version pinning, all with only a simple JSON config file and apparently no need for being/becoming a Nix expert. There's no free lunch, but Nix is more than a dev environment solution, so we probably don't actually need its full power for this purpose. Also, aren't we mixing the two uses of Nix at the moment? I wonder what you think about Devbox, @yajo. From my non-expert perspective, it looks quite appealing.

/cc @pawamoy

**Comment by pawamoy** [comment]
No strong opinion from me. Devbox looks cool, but I didn't take time to actually try it. I don't work a lot on Copier, but had a few issues with its Nix setup indeed. Anything that makes Copier's dev env easier to use and reason about is welcome :slightly_smiling_face:

**Reply by pawamoy** [reply]
So I've played with Devbox a bit, and I like it. It's very simple to use. I'll probably add support for it in my own projects. Strong emphasis on *support*: I don't want to make it the default way to setup the dev env. I think Copier's contributing instructions should do the same and provide instructions to setup the dev env *without* nix and direnv. Currently, instructions ask to install direnv, which is easy enough, but requires the user to load direnv into their shell environment (not cool if they're not a direnv user already). I guess direnv then uses the flake.nix stuff to install nix and the rest of the dependencies, though I personally had to also enable the nix service with `sudo systemctl enable nix.service`, which again feels heavy for non-users. For users who simply *cannot* install direnv and/or nix, there is no instructions explaining what to install and how (Poetry and Python deps).

(Side note: when using direnv and nix, I get warnings I don't understand:)

```console
% cd copier
warning: ignoring untrusted substituter 'https://copier.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
warning: ignoring untrusted substituter 'https://devenv.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
pre-commit-hooks.nix: hooks up to date

% poe lint
Poe => lint
warning: ignoring untrusted substituter 'https://copier.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
warning: ignoring untrusted substituter 'https://devenv.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
warning: ignoring untrusted substituter 'https://copier.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
warning: ignoring untrusted substituter 'https://devenv.cachix.org', you are not a trusted user.
Run `man nix.conf` for more information on the `substituters` configuration option.
pre-commit-hooks.nix: hooks up to date
alejandra................................................................Passed
editorconfig-checker.....................................................Passed
prettier.................................................................Passed
ruff.....................................................................Passed
taplo....................................................................Passed
```

**Reply by yajo** [reply]
The main problem is pre-commit. I don't want to use it to install stuff because it gives problems all the time. Also, we cannot rely on Poetry to install formatters because we'd still need a solution for taplo and Prettier.

However, we could have an action that pushes formatter changes back to the PRs, so people without nix can just push and wait. WDYT?

**Comment by yajo** [comment]
I've been reading over it, but TBH it doesn't seem to add a lot if you're used to Nix, which I happen to be.

I think that Nix, as a technology, is very awesome and the industry need for something like it is big. However, it is scary at the beginning, so it's just a matter of time that some projects pop up here and there simplifying the 1st encounters with it, just like this one.

There's a similar project to the one you pointed out, https://devenv.sh/, which we are actually using in Copier. I just didn't go down its yaml path because I understand Nix and feel more comfortable writing it than YAML. But if you feel it would be less scary, I can translate some of its configs to YAML.

The setup is rather simple right now. All you need is:
1. Install nix (you need this in devbox too).
2. Run ˋnix developˋ.

You can actually run linters without installing nix if you have podman or docker and run ˋpoe lintˋ.

What are the pain points you encounter with this? 🤔

**Reply by sisp** [reply]
Sorry for my delayed reply.

My biggest pain points are readability of the Nix language and specifying dependency versions explicitly (and easily) for each dev dependency instead of relying on the versions implied by the Nix distribution. For instance, it's not so clear which Ruff version we're currently using. At first glance, it looks very straightforward with Devbox.

**Reply by yajo** [reply]
Regarding readability, would you feel more comfortable if I rewrite some configs to YAML?

Regarding versions, you just need to run `ruff --version`. IMHO it isn't so important to be able to choose a specific linter version. You'll simply want the latest stable almost always. It's more important the fact that we know it works and it's reproducible.

**Reply by danieleades** [reply]
the current version of ruff in the nix environment is currently 0.1.6, whereas the latest available version is 0.2.2, so it's lagging behind quite a few releases.

I see a couple of couple of other issues-

- the flake8 config is totally redundant and contains no lint groups not provided by ruff (except the `C` lint group, which i'm guessing is meant to be the `C90` complexity lint group. Ruff supports this but it's not currently enabled). Flake8 can be removed entirely as far as i can tell
- formatting doesn't appear to be checked. There's a `ruff check` pre-commit target, but no `ruff format` pre-commit target
- is there a mypy check somewhere or is it just installed into the dev environment?

i've never worked with Nix, but all of this is utterly opaque to me.

**Comment by yajo** [comment]
Mypy is checked here:
https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/tests/test_tools.py#L14

Since pre-commit is handled by nix, it is also tested by nix:
https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/.github/workflows/ci.yml#L128

That line builds all derivations found inside the ˋchecksˋ nix flake
output. One of those is ˋdevenv-ciˋ, which is provided by devenv.sh (it's
called just ˋciˋ there) and inherited here:
https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/flake.nix#L132

More docs on devenv pre-commit integration:
https://devenv.sh/pre-commit-hooks/

The reason for having such old versions is that I don't have a job that
auto-updates [the nix lock](
https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/flake.lock).
I'll install https://github.com/marketplace/actions/update-nix-flake-lock
when I have time (it didn't exist when I included nix support).

Check here how nix totaly surpasses any other distro you can imagine in
amount of packages available and updated:
https://repology.org/repositories/statistics/newest

That is just because one simple reason: creating and maintaining packages
with nix is extremely simple and maintainable. But of course, you have to
know Nix. Just like you need to know Python to write Copier! :D

In any case, I don't expect Copier contributors to know how to write Nix.
It's a science apart and it'd be a big contribution barrier. I just want a
smooth as possible contribution experience. These days, if you don't have
Nix installed but you happen to have podman or docker installed, just run
ˋpoe lintˋ and it will just work.

However I'll try to improve the experience and docs just to make it
smoother for people that care about the gory details. For some reason,
Copier attracts a lot of those people (and I love that 🤎)!

**Reply by sisp** [reply]
> The reason for having such old versions is that I don't have a job that auto-updates [the nix lock]( https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/flake.lock). I'll install https://github.com/marketplace/actions/update-nix-flake-lock when I have time (it didn't exist when I included nix support).

Would https://docs.renovatebot.com/modules/manager/nix/ seem to help as well? I have gained some Renovate experience recently and might give migrating Copier from Dependabot to Renovate a try (https://github.com/orgs/copier-org/discussions/1242#discussioncomment-6460096). This would also help with updating transitive dependencies in `poetry.lock`.

**Reply by yajo** [reply]
Yes of course. That hook actually updates this line (which I have to update twice a year by hand): https://github.com/copier-org/copier/blob/4ff30e8a98596a580bd5f8d788657204fca7fbf4/flake.nix#L23

According to the page you linked, ˋflake.lockˋ maintenance is done with another configuration.

But in any case, go ahead please. I'd love to see how renovate works, compared to dependabot.
